docs/edsl.md

```diff
@@ -0,0 +1,735 @@
+same line as contractions, so we write the global min function (for a 3D tensor)
+as follows:
+
+```c++
+Tensor global_min(const Tensor& I) {
+  TensorIndex i, j, k;
+  auto Neg = -I;
+  auto O_Neg = TensorOutput();
+  O_Neg() >= Neg(i, j, k);
+  auto O = -O_Neg;
+  return O;
+}
+```
+
+There are several novel pieces in this example. First, note that the elementwise
+operations do not include dimensions. Dimensions are inferred from the inputs in
+elementwise operations, and so are never specified in elementwise ops. `Neg` has
+the same shape as `I`, and `O` has the same shape as `O_Neg`. When an
+elementwise binary operation is performed, the output shape is determined using
+[broadcasting semantics][].
+
+Which brings us to the next novelty: we have our first example of a 0D tensor,
+`O_Neg`. Tensors in Tile are allowed to have zero dimensions. In such a case the
+tensor represents a scalar, i.e., a single value. In places where dimensions are
+specified, you can indicate a 0-dimensional tensor by using `()` for the
+dimensions, as in this example.
+
+Notice that we are taking the max over all axes in a single operation.
+Contractions implicitly aggregate over _all_ indices that write to the same
+output location (in this case we aggregate over all values of `i`, `j`, and
+`k`).
+
+### Average
+
+To compute the mean of a tensor, we need to sum the elements and divide by the
+total number of elements summed. We can do this by taking advantage of the fact
+that we can divide by a constant (including an input `TensorDim`) as an
+elementwise operation. Thus, to take the mean over axis `0` of a 2D tensor, we
+write:
+
+```c++
+Tensor avg(const Tensor& I) {
+  TensorDim X, Y;
+  TensorIndex x, y;
+  I.bind_dims(X, Y);
+  auto Sum = TensorOutput();
+  Sum(y) += I(x, y);
+  return Sum / X;
+}
+```
+
+We can perform multiple elementwise operations on the same line, including
+operations on constants and input dimensions. So, while it would be possible to
+take a global mean of a 2D tensor in stages as so:
+
+```
+Tensor avg(const Tensor& I) {
+  TensorDim X, Y;
+  TensorIndex x, y;
+  I.bind_dims(X, Y);
+  auto Sum = TensorOutput();
+  Sum() += I(x, y);
+  PartialMean = Sum / X;
+  return PartialMean / Y;
+}
+```
+
+it is more straightforward to merge the elementwise operations:
+
+```
+Tensor avg(const Tensor& I) {
+  TensorDim X, Y;
+  TensorIndex x, y;
+  I.bind_dims(X, Y);
+  auto Sum = TensorOutput();
+  Sum() += I(x, y);
+  return Sum / (X * Y);
+}
+```
+
+### Max Pool 1D
+
+Next let's implement a size 2 stride 2 maxpool in Tile. This is the operation
+that splits a tensor into groups of 2 and takes the larger element from each
+group, yielding a tensor of half the original size. This is straightforward to
+implement in straight C++:
+
+```c++
+float I[N], O[N / 2];
+for (int i = 0; i < N/2; ++i) {
+    float curr_max = FLT_MIN;
+    for (int j = 0; j < 2; ++j) {
+      curr_max = I[2 * i + j];
+      O.add_constraints({I[2 * i + j] > curr_max})
```

Comment:
So, this snippet should continue to use the `if` because this is an example of how we'd write it in straight C++.